Document same-role permissions for group-based access control (#502)
mautic/developer-documentation-new: docs/plugin_services/security.rst

   * - ``addStandardPermissions()``
     - Set view, edit, create, delete, publish - with option to exclude, and full permissions.
   * - ``addExtendedPermissions()``

>      - Set creator level restrictions: ``viewown``, ``viewsamerole``, ``viewother``, ``editown``, ``editsamerole``, ``editother``, ``create``, ``deleteown``, ``deletesamerole``, ``deleteother``, ``publishown`` - with option to exclude, ``publishsamerole`` - with option to exclude, ``publishother`` - with option to exclude, and ``full``. The ``samerole`` permissions allow users to access content created by other users who share the same Role.

Reviewer: ```suggestion
     - Set creator level restrictions: ``viewown``, ``viewsamerole``, ``viewother``, ``editown``, ``editsamerole``, ``editother``, ``create``, ``deleteown``, ``deletesamerole``, ``deleteother``, ``publishown`` - with option to exclude, ``publishsamerole`` - with option to exclude, ``publishother`` - with option to exclude, and ``full``. The ``samerole`` permissions allow Users to access content created by other Users who share the same Role.
```

   * - ``addStandardFormFields()``
     - Appends the standard permission sets to the Form
   * - ``addExtendedFormFields()``

>      - Appends the extended, aka creator restricted, permissions to the Form, including same-role permissions 

Reviewer: ```suggestion
     - Appends the extended, aka creator restricted, permissions to the Form, including Same Role permissions 
```

The same applies for the method ``isSupported()`` which you can use to determine if a bundle or Plugin includes the requested permission and permission level.
You can also use this to provide BC support.

Same-role permissions

> *********************

Reviewer: ```suggestion
.. vale off

Same Role permissions
*********************

.. vale on
```

> 

Author: PR #15465 introduces the same-role permissions feature. The new bit values (2048, 4096, 8192, 16384) are from `AbstractPermissions::addExtendedPermissions()` at lines 474-793 of the diff.

Source: https://github.com/mautic/mautic/pull/15465

> Same-role permissions provide a middle ground between ``own`` and ``other`` permissions. When a user has a same-role permission such as ``viewsamerole``, they can access content created by any user who has the same Role assigned to them.

Reviewer: ```suggestion
Same Role permissions provide a middle ground between ``own`` and ``other`` permissions. When a User has a Same Role permission such as ``viewsamerole``, they can access content created by any User who has the same Role assigned to them.
```

Available same-role permissions

> ===============================

Reviewer: ```suggestion
.. vale off

Available Same Role permissions
===============================

.. vale on
```

> When using ``addExtendedPermissions()``, the following same-role permissions become available:

Reviewer: ```suggestion
When using ``addExtendedPermissions()``, the following Same Role permissions become available:
```

.. list-table::
   :header-rows: 1

   * - Permission
     - Description
     - Bit
   * - ``viewsamerole``

>      - View content created by users with the same Role

Reviewer: ```suggestion
     - View content created by Users with the same Role
```

     - 2048
   * - ``editsamerole``

>      - Edit content created by users with the same Role

Reviewer: ```suggestion
     - Edit content created by Users with the same Role
```

     - 4096
   * - ``deletesamerole``

>      - Delete content created by users with the same Role

Reviewer: ```suggestion
     - Delete content created by Users with the same Role
```

     - 8192
   * - ``publishsamerole``

>      - Publish/unpublish content created by users with the same Role

Reviewer: ```suggestion
     - Publish/unpublish content created by Users with the same Role
```

     - 16384

Permission hierarchy
====================

The permission hierarchy from most restrictive to least restrictive is:

1. ``own`` - Access only content created by the current user
2. ``samerole`` - Access content created by users with the same Role

> 3. ``other`` - Access content created by any user

Reviewer: Use Mautic style guide.

```suggestion
#. ``own`` - Access only content created by the current User
#. ``samerole`` - Access content created by Users with the same Role
#. ``other`` - Access content created by any User
```

> When checking access, Mautic evaluates permissions in this order. If a user has ``viewother``, they can view all content regardless of same-role settings.

Reviewer: ```suggestion
When checking access, Mautic evaluates permissions in this order. If a User has ``viewother``, they can view all content regardless of Same Role settings.
```

Checking entity access with roles

> =================================

Reviewer: ```suggestion
Checking entity access with Roles
=================================
```

> The ``hasEntityAccess()`` method supports same-role permissions. When you pass ``viewown`` and ``viewother`` permissions, Mautic automatically infers the corresponding ``viewsamerole`` permission:

Reviewer: ```suggestion
The ``hasEntityAccess()`` method supports Same Role permissions. When you pass ``viewown`` and ``viewother`` permissions, Mautic automatically infers the corresponding ``viewsamerole`` permission:
```

.. code-block:: php

   <?php

> 

Author: The `hasEntityAccess()` method signature and auto-inference of same-role permissions is from `CorePermissions` at lines 870-1012 of the PR diff, specifically the `toSameRolePermission()` private method.

Source: https://github.com/mautic/mautic/pull/15465

>    // Mautic automatically checks viewsamerole when viewown/viewother are passed

Reviewer: There are a couple of things:

- Use active instead of passive voice for "are passed"
- Use other word than "checks"

   if ($this->security->hasEntityAccess(
       'email:emails:viewown',
       'email:emails:viewother',
       $entity->getCreatedBy()
   )) {
       // User has access via own, same-role, or other permission
   }

> You can also explicitly pass a same-role permission as the fourth parameter:

Reviewer: ```suggestion
You can also explicitly pass a Same Role permission as the fourth parameter:
```

.. code-block:: php

   <?php

   if ($this->security->hasEntityAccess(
       'lead:lists:viewown',
       'lead:lists:viewother',
       $entity->getCreatedBy(),
       'lead:lists:viewsamerole'
   )) {
       // Explicit same-role permission check
   }

Controller helpers for role-based filtering

> ===========================================

Reviewer: ```suggestion
.. vale off

Controller helpers for Role-based filtering
===========================================

.. vale on
```

> Controllers extending ``CommonController`` have access to helper methods that handle role-based filtering of entity lists:

Reviewer: ```suggestion
Controllers extending ``CommonController`` have access to helper methods that handle Role-based filtering of entity lists:
```

> .. code-block:: php

Author: Controller helper methods `getStandardPermissions()`, `hasStandardViewPermission()`, and `addStandardRoleBasedFilter()` are from `CommonController` at lines 490-547 of the PR diff.

Source: https://github.com/mautic/mautic/pull/15465

   <?php

>    // Get standard permissions including same-role permissions

Reviewer: ```suggestion
   // Get standard permissions including Same Role permissions
```

   $permissionBase = 'email:emails';
   $permissions = $this->getStandardPermissions($permissionBase);

>    // Check if user has any view permission (own, same-role, or other)

Reviewer: There are a couple of things here:

- Use other word than "check"
- Capitalize the permissions and use dash instead of parantheses

```suggestion
   // Check if user has any view permission - Own, Same Role, or other
```

   if (!$this->hasStandardViewPermission($permissions, $permissionBase)) {
       return $this->accessDenied();
   }

>    // Add role-based filtering to entity queries

Reviewer: ```suggestion
   // Add Role-based filtering to entity queries
```

   $filter = ['string' => $search, 'force' => []];
   $this->addStandardRoleBasedFilter($filter, $permissions, $permissionBase, 'e.createdBy');

The ``addStandardRoleBasedFilter()`` method:

- Does nothing if the user has ``viewother`` permission
- Filters to users with the same Role if the user has ``viewsamerole`` permission

> - Filters to only the current user if neither ``viewother`` nor ``viewsamerole`` is granted

Reviewer: There are a couple of things here:

- Use Mautic style guide for unordered list
- Use active instead of passive voice for "is granted" at the last bullet point.

```suggestion
* Does nothing if the User has ``viewother`` permission
* Filters to Users with the same Role if the User has ``viewsamerole`` permission
* Filters to only the current User if neither ``viewother`` nor ``viewsamerole`` is granted
```

The same applies for the method ``isSupported()`` which you can use to determine if a bundle or Plugin includes the requested permission and permission level.
You can also use this to provide BC support.

.. vale off

Same Role permissions
*********************

.. vale on

Same Role permissions provide a middle ground between ``own`` and ``other`` permissions. A Same Role permission such as ``viewsamerole`` allows a User to access content created by any User who has the same Role assigned to them.

.. vale off

Available Same Role permissions
===============================

.. vale on

When using ``addExtendedPermissions()``, the following Same Role permissions become available:

.. list-table::
   :header-rows: 1

   * - Permission
     - Description
     - Bit
   * - ``viewsamerole``
     - View content created by Users with the same Role
     - 2048
   * - ``editsamerole``
     - Edit content created by Users with the same Role
     - 4096
   * - ``deletesamerole``
     - Delete content created by Users with the same Role
     - 8192
   * - ``publishsamerole``

>      - Publish/Unpublish content created by Users with the same Role

Reviewer: ```suggestion
     - Activate/Deactivate content created by Users with the same Role
```

>      - Publish/unpublish content created by Users with the same Role

Reviewer: ```suggestion
     - Publish/Unpublish content created by Users with the same Role
```

     - 16384

Permission hierarchy
====================

The permission hierarchy from most restrictive to least restrictive is:

#. ``own`` - Access only content created by the current User
#. ``samerole`` - Access content created by Users with the same Role
#. ``other`` - Access content created by any User

> When checking access, Mautic evaluates permissions in this order. If a User has ``viewother``, they can view all content regardless of Same Role settings.

Reviewer: Use other word than 'checking'.

Checking entity access with Roles

> =================================

Reviewer: Use other word than 'checking', and wrap this in vale off/on statements.

```suggestion
.. vale off

Checking entity access with Roles
=================================

.. vale on
```

       'lead:lists:viewown',
       'lead:lists:viewother',
       $entity->getCreatedBy(),
       'lead:lists:viewsamerole'
   )) {
       // Explicit same-role permission check
   }

.. vale off

Controller helpers for Role-based filtering
===========================================

.. vale on

Controllers extending ``CommonController`` have access to helper methods that handle Role-based filtering of entity lists:

.. code-block:: php

   <?php

   // Get standard permissions including Same Role permissions
   $permissionBase = 'email:emails';
   $permissions = $this->getStandardPermissions($permissionBase);

   // Check if user has any view permission - Own, Same Role, or other
   if (!$this->hasStandardViewPermission($permissions, $permissionBase)) {
       return $this->accessDenied();
   }

   // Add Role-based filtering to entity queries
   $filter = ['string' => $search, 'force' => []];
   $this->addStandardRoleBasedFilter($filter, $permissions, $permissionBase, 'e.createdBy');

The ``addStandardRoleBasedFilter()`` method:

* Does nothing if the User has ``viewother`` permission
* Filters to Users with the same Role if the User has ``viewsamerole`` permission

> * Filters to only the current User if neither ``viewother`` nor ``viewsamerole`` is granted

Reviewer: Use active instead passive voice for "is granted"

Reviewer: @promptless-for-oss please address this.
